subprojects/frameworklauncher/yarn/src/main/java/com/microsoft/frameworklauncher/applicationmaster/FrameworkInfoPublisher.java

```diff
@@ -82,9 +82,9 @@ protected void run() throws Exception {
     new Thread(() -> {
       while (true) {
         try {
-          publishFrameworkInfo();
-
           Thread.sleep(conf.getLauncherConfig().getAmFrameworkInfoPublishIntervalSec() * 1000);
+
+          publishFrameworkInfo();
```

Comment:
Why move it?

Reply:
@yqwang-ms, this is the bug with infinite spin-waiting you were asking about. If `publishFrameworkInfo()` has thrown, the timeout was never started. This function throws, at least, if `zkStore` is unavailable. It becomes unavailable, for example, after its stop in the stop signal handling thread.

Reply:
But why zkStore can be unavailable? When we stop AM, we stop the zkStore at last.

Reply:
Yes, it happened during the stopping as I described above. The impact is ~1-10 Gb of useless error spam in AM logs during the stop.

Reply:
If so, maybe we can skip to call zkStore.stop() (it is the old behaiour and works fine)

Reply:
Should be done in another PR.

Reply:
@zhiltsov-max, to unblock the launcher fix, I have moved it to another PR:
#2435
Thanks for reporting and the proposal solution!